src/csync/vio/csync_vio_local_unix.cpp

```diff
@@ -99,13 +99,13 @@ std::unique_ptr<csync_file_stat_t> csync_vio_local_readdir(csync_vio_handle_t *d
   }
 
   file_stat.reset(new csync_file_stat_t);
-  file_stat->path = c_utf8_from_locale(dirent->d_name);
-  if (file_stat->path.isNull()) {
-      file_stat->original_path = handle->path;
-      file_stat->original_path += '/';
-      file_stat->original_path += dirent->d_name;
-      CSYNC_LOG(CSYNC_LOG_PRIORITY_WARN, "Invalid characters in file/directory name, please rename: \"%s\" (%s)",
+  if (*dirent->d_name) {
+    file_stat->path = c_utf8_from_locale(dirent->d_name);
+    file_stat->original_path = QByteArray() % const_cast<const char *>(handle->path) % '/' % const_cast<const char *>(dirent->d_name);
```

Comment:
This keeps more memory around (increases the peak) and makes unix behave differently than Windows.

Do we need to separate `stat` from `readdir` for the local tree (and we don't stat at all for the remote tree). Could we merge them in one call instead (or make `readdir` call `stat` itself with the right `mbchar_t`)? This would make Windows also benefit from the improvement, and prevent reading from `buf` which was write-only until now.

Reply:
> Do we need to separate stat from readdir for the local tree (since we don't stat at all for the remote tree)? Could we merge them in one call instead (or make readdir call stat itself with the right mbchar_t)? 


Related discussion: https://github.com/owncloud/client/issues/880

Would it make sense to switch directly to ```QDirIterator```?